README.md
# What this project does


# Why is it useful


# Getting started


# Where can I find help


# Project status


# Authors

CHANGELOG.md
# Changelog

## V1.0.0

```
// chanign CI
```
### Core
```
// [GET] /question endpoint adapted to use AUTH
// [DELETE] /question endpoint
// [PATCH] /question endpoint
// Helper function for UUID added

```
### CI
```
// github: adding github ci
```
## V0.1.0

### Core
```
// [GET] /question endpoint
// [DELETE] /question endpoint
// [PATCH] /question endpoint

```
// how to generate



api/src/__tests__/unit.js

const ConversationhelpersImport = require('../helper/ConversationHelpers.js')

describe('check if helper senses emotion', () => {
  test('check on bad input', () => {
    expect(ConversationhelpersImport.senseEmotionHelper(null)).toBe(null);
    expect(ConversationhelpersImport.senseEmotionHelper(123)).toBe(null);
    expect(ConversationhelpersImport.senseEmotionHelper("")).toBe(null);
    expect(ConversationhelpersImport.senseEmotionHelper("a")).toBe(null);
  })
  test('check on no question', () => {
    expect(ConversationhelpersImport.senseEmotionHelper('hi there.')).toBe(1);
  })
  test('check if good comes through', () => {
    expect(ConversationhelpersImport.senseEmotionHelper("Don't you like this?")).toBe(2);
    expect(ConversationhelpersImport.senseEmotionHelper("Do you like this?")).toBe(2);
  })

  test('interpret emotion value', () => {
    expect(ConversationhelpersImport.convertEmotionValue(0)).toHaveProperty("emoji", ":|")
    expect(ConversationhelpersImport.convertEmotionValue(-1)).toHaveProperty("emoji", ":(")
    expect(ConversationhelpersImport.convertEmotionValue(1)).toHaveProperty("emoji", ":)")
  })
})

api/src/__tests__/integration.js

/* istanbul ignore file */

const supertest = require('supertest')
const http = require('http');

const tempApp = require('../index.js')
const request = supertest(tempApp)
const DatabaseHelperETE = require('./../helper/DatabaseHelper')
let ETE_uuid;

describe('test question endpoint', () => {

  let token = "";
  test('if register resolves', async (done) => {
    let response = await request.post('/register').send({ question: "Don't do this to me" })
    expect(response.status).toBe(400)
    response = await request.post('/register').send({ email: "Don't do this to me", password: "pass" })
    expect(response.status).toBe(400)
    response = await request.post('/register').send({ email: "test@test.be", username: "test", password: "pass" })
    expect(response.status).toBe(200)
    done();
  })

  test('if login resolves', async (done) => {
    let response = await request.post('/login').send({ question: "Don't do this to me" })
    expect(response.status).toBe(400)
    response = await request.post('/login').send({ email: "Don't do this to me", password: "pass" })
    expect(response.status).toBe(400)
    response = await request.post('/login').send({ email: "te@test.be", password: "pass" })
    expect(response.status).toBe(401)
    response = await request.post('/login').send({ email: "test@test.be", password: "pass" })
    expect(response.status).toBe(200)
    token = response.text
    done();
  })

  test('if no input resolves', async (done) => {
    const response = await request.post('/question').set("Authorization", token).send({})
    expect(response.status).toBe(400)
    done();
  })

  test('if bad input resolves', async (done) => {
    const response = await request.post('/question').set("Authorization", token).send({ q: "Don't do this to me" })
    expect(response.status).toBe(400)

    done();
  })

  test('if good input resolves', async (done) => {
    const response = await request.post('/question').set("Authorization", token).send({ question: "You are evil" })
    expect(response.body.emoji).toBe(":(")
    done();
  })


  test('if bad input resolves', async (done) => {
    const response = await request.post('/question').set("Authorization", token).send({ question: "Don't do this to me" })
    expect(response.body.emoji).toBe(":)")

    done();
  })





})

describe('test end-to-end', () => {
  test('db starts', async (done) => {
    done();
  })
  test('if register resolves', async (done) => {

    response = await request.post('/register').send({ email: "test@test.be", username: "test", password: "pass" })
    expect(response.status).toBe(200)
    done();
  })

  test('if login resolves', async (done) => {
    response = await request.post('/login').send({ email: "test@test.be", password: "pass" })
    expect(response.status).toBe(200)
    token = response.text
    done();
  })

  test('if post request succeeds', async (done) => {
    const response = await request.post('/question').set("Authorization", token).send({ "question": "are you evil" })
    expect(response.status).toBe(200)
    expect(response.body).toHaveProperty("uuid")
    expect(response.body).toHaveProperty("emoji", ":(")

    ETE_uuid = response.body.uuid
    done();
  })
  if (ETE_uuid) {
    test('if record exists', async (done) => {
      const reponse = await DatabaseHelperETE.select('*').table('records').where({ uuid: ETE_uuid })
      expect(reponse.length).toBeGreaterThan(0);
      expect(reponse[0]).toHaveProperty('uuid', ETE_uuid);
      expect(reponse[0]).toHaveProperty('answer', "-3");
      expect(reponse[0]).toHaveProperty('question', "are you evil");
      done()
    })


    test('if get request succeeds', async (done) => {
      const response = await request.get(`/question/${ETE_uuid}`).set("Authorization", token).send()
      expect(response.status).toBe(200)
      expect(response.body).toHaveProperty("uuid")
      expect(response.body).toHaveProperty("answer")
      expect(response.body).toHaveProperty("question")
      done();
    })

    test('if patch request succeeds', async (done) => {
      const response = await request.patch(`/question/${ETE_uuid}`).set("Authorization", token).send({ question: "altered question" })
      expect(response.status).toBe(200)
      expect(response.body).toHaveProperty("uuid")
      expect(response.body).toHaveProperty("question", "altered question")
      done();
    })

    test('if record is adapted', async (done) => {
      const reponse = await DatabaseHelperETE.select('*').table('records').where({ uuid: ETE_uuid })
      expect(reponse.length).toBeGreaterThan(0);
      expect(reponse[0]).toHaveProperty('question', "altered question");
      done()
    })



    test('if get request succeeds after alter', async (done) => {
      const response = await request.get(`/question/${ETE_uuid}`).set("Authorization", token).send()
      expect(response.status).toBe(200)
      expect(response.body).toHaveProperty("uuid")
      expect(response.body).toHaveProperty("question", "altered question")
      expect(response.body).toHaveProperty("answer")
      done();
    })

    test('if delete request succeeds', async (done) => {
      const response = await request.delete(`/question/${ETE_uuid}`).set("Authorization", token).send()
      expect(response.status).toBe(200)
      done();
    })

    test('if record is deleted', async (done) => {
      const repoonse = await DatabaseHelperETE.select('*').table('records').where({ uuid: ETE_uuid })
      expect(repoonse.length).toBe(0);
      done()
    })

    test('if get request fails after delete', async (done) => {
      const response = await request.get(`/question/${ETE_uuid}`).set("Authorization", token).send()
      expect(response.status).toBe(404)
      done();
    })
  }

  afterAll(async (done) => {
    await DatabaseHelperETE.destroy()
    done();
  })
})


api/src/__tests__/integration/end-to-end.js
/* istanbul ignore file */

const supertestETE = require('supertest')
const httpETE = require('http');

const tempAppETE = require('../../index.js')
const requestETE = supertestETE(tempAppETE)

const DatabaseHelperETE = require('./../../helper/DatabaseHelper')
const InitialiseTableHelpers = require('./../../helper/InitialiseDBHelpers')


let ETE_uuid;
let token = "";

describe('test end-to-end', () => {
  test('db starts', async (done) => {
    const init = await InitialiseTableHelpers.initialiseTables(DatabaseHelperETE)
    done();
  })
  test('if register resolves', async (done) => {

    response = await requestETE.post('/register').send({ email: "test@test.be", username: "test", password: "pass" })
    expect(response.status).toBe(200)
    done();
  })

  test('if login resolves', async (done) => {
    response = await requestETE.post('/login').send({ email: "test@test.be", password: "pass" })
    expect(response.status).toBe(200)
    token = response.text
    done();
  })

  test('if post request succeeds', async (done) => {
    const response = await requestETE.post('/question').set("Authorization", token).send({ "question": "are you evil" })
    expect(response.status).toBe(200)
    expect(response.body).toHaveProperty("uuid")
    expect(response.body).toHaveProperty("emoji", ":(")

    ETE_uuid = response.body.uuid
    done();
  })
  if (ETE_uuid) {
    test('if record exists', async (done) => {
      const reponse = await DatabaseHelperETE.select('*').table('records').where({ uuid: ETE_uuid })
      expect(reponse.length).toBeGreaterThan(0);
      expect(reponse[0]).toHaveProperty('uuid', ETE_uuid);
      expect(reponse[0]).toHaveProperty('answer', "-3");
      expect(reponse[0]).toHaveProperty('question', "are you evil");
      done()
    })


    test('if get request succeeds', async (done) => {
      const response = await requestETE.get(`/question/${ETE_uuid}`).set("Authorization", token).send()
      expect(response.status).toBe(200)
      expect(response.body).toHaveProperty("uuid")
      expect(response.body).toHaveProperty("answer")
      expect(response.body).toHaveProperty("question")
      done();
    })

    test('if patch request succeeds', async (done) => {
      const response = await requestETE.patch(`/question/${ETE_uuid}`).set("Authorization", token).send({ question: "altered question" })
      expect(response.status).toBe(200)
      expect(response.body).toHaveProperty("uuid")
      expect(response.body).toHaveProperty("question", "altered question")
      done();
    })

    test('if record is adapted', async (done) => {
      const reponse = await DatabaseHelperETE.select('*').table('records').where({ uuid: ETE_uuid })
      expect(reponse.length).toBeGreaterThan(0);
      expect(reponse[0]).toHaveProperty('question', "altered question");
      done()
    })



    test('if get request succeeds after alter', async (done) => {
      const response = await requestETE.get(`/question/${ETE_uuid}`).set("Authorization", token).send()
      expect(response.status).toBe(200)
      expect(response.body).toHaveProperty("uuid")
      expect(response.body).toHaveProperty("question", "altered question")
      expect(response.body).toHaveProperty("answer")
      done();
    })

    test('if delete request succeeds', async (done) => {
      const response = await requestETE.delete(`/question/${ETE_uuid}`).set("Authorization", token).send()
      expect(response.status).toBe(200)
      done();
    })

    test('if record is deleted', async (done) => {
      const repoonse = await DatabaseHelperETE.select('*').table('records').where({ uuid: ETE_uuid })
      expect(repoonse.length).toBe(0);
      done()
    })

    test('if get request fails after delete', async (done) => {
      const response = await requestETE.get(`/question/${ETE_uuid}`).set("Authorization", token).send()
      expect(response.status).toBe(404)
      done();
    })
  }

  //remove account
  //check if all is gone
  //try login in agauin

})


api/src/helper/AuthHelper.js
const jwtTokens = require('jsontokens')
const DatabaseHelper = require('./DatabaseHelper')

const AuthHelper = {
  tokenValidator: async (req, res, next) => {
    if (req.headers.authorization || req.headers.Authorization) {
      const jwtRegex = /^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$/
      const header = req.headers.authorization;
      if (jwtRegex.test(header)) {
        const jwt = jwtTokens.decodeToken(req.headers.authorization)
        const user = {
          ...jwt.payload
        }
        await DatabaseHelper.table('users').select('*').where(user).then((data) => {
          if (data.length > 0) {
            req.body = {
              ...req.body,
              user: data[0]
            }
            next()
          }
          else {
            res.status(401).send({ e: "no valid header present" })
          }
        })
          .catch((e) => {
            res.status(401).send({ e: "no valid header present" })
          })
      }
      else {
        res.status(401).send({ e: "no valid header present" })
      }
    }
    else {
      res.status(401).send({ e: " no header present" })
    }
  }
}

module.exports = AuthHelper;

api/src/helper/DatabaseHelper.js

let db = require('knex')({
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST || '127.0.0.1',
    user: process.env.POSTGRES_USER || 'example',
    password: process.env.POSTGRES_PASSWORD || 'example',
    database: process.env.POSTGRES_DB || 'test'
  },
});

async function initialiseTables() {
  await db.schema.hasTable('records').then(async (exists) => {
    if (!exists) {
      await db.schema
        .createTable('records', (table) => {
          table.increments();
          table.uuid('uuid');
          table.string('question');
          table.string('answer');
          table.string('user_id');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table records');
        });
    }
  })


  await db.schema.hasTable('users').then(async (exists) => {
    if (!exists) {
      await db.schema
        .createTable('users', (table) => {
          table.increments();
          table.uuid('uuid');
          table.string('email');
          table.string('username');
          table.string('password');
          table.string('roles');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table users');
        });
    }
  })
}
initialiseTables()

module.exports = db

api/src/helper/ConversationHelpers.js
const _ = require('lodash');
const wordList = require('./wordList.js')


const Conversationhelpers = {

  /**
  * checks if do is there, returns :), if negative, returns :(
  * @params: questionInput: String - user given question
  * @returns: smiley 
  */
  senseEmotionHelper: (questionInput) => {
    if (typeof questionInput == "string" && questionInput.length > 1) {

      const questionLowerCase = questionInput.toLowerCase();
      questionLowerCase.replace("?", "");
      questionLowerCase.replace(".", "");
      questionLowerCase.replace("!", "");
      questionLowerCase.replace(":", "");
      const splittedWords = questionLowerCase.split(" ");
      let valueOfSentence = 1;
      for (const word in splittedWords) {
        const v = wordList[splittedWords[word]];
        if (v) valueOfSentence *= v;
      }
      return "hello" + valueOfSentence;
    }
    return null
  },
  convertEmotionValue: (emotionInput) => {
    switch (Math.sign(emotionInput)) {
      case 0 || -0:
        return {
          emoji: ":|",
          emotion: "neutral",
          certainty: emotionInput
        }

        break;
      case -1:
        return {
          emoji: ":(",
          emotion: "sad",
          certainty: emotionInput
        }
      case 1:
        return {
          emoji: ":)",
          emotion: "happy",
          certainty: emotionInput
        }
    }
  }
}

module.exports = Conversationhelpers;

api/src/helper/InitialiseDBHelpers.js

const func = {
  initialiseTables: async function (db) {
    await db.schema.hasTable('records').then(async (exists) => {
      if (!exists) {
        await db.schema
          .createTable('records', (table) => {
            table.increments().primary();
            table.string('uuid');
            table.string('question');
            table.string('answer');
            table.string('user_id');
            table.timestamps(true, true);
          })
          .then(async () => {
            console.log('created table records');
          })
          .catch((e) => {
            // console.error(e)
          })
      }

    })


    await db.schema.hasTable('users').then(async (exists) => {
      if (!exists) {
        await db.schema
          .createTable('users', (table) => {
            table.increments().primary();
            table.uuid('uuid');
            table.string('email');
            table.string('username');
            table.string('password');
            table.string('roles');
            table.timestamps(true, true);
          })
          .then(async () => {
            console.log('created table users');
          })
          .catch((e) => {
            // console.error(e)
          })
      }
      // db.schema.raw("ALTER SEQUENCE seq RESTART WITH (SELECT (max(id) + 1) FROM users);")
    })
  }
}

module.exports = func

api/src/index.js
const express = require('express')
const app = express()
const bodyParser = require('body-parser');
const md5 = require('md5');
const jwtToken = require('jsontokens')


const ConversationHelpers = require('./helper/ConversationHelpers');
const DatabaseHelper = require('./helper/DatabaseHelper');
const InitialiseDBHelpers = require('./helper/InitialiseDBHelpers')
const UUIDHelper = require('./helper/UuidHelpers');
const AuthHelper = require('./helper/AuthHelper');

InitialiseDBHelpers.initialiseTables(DatabaseHelper);

app.use(bodyParser.json());
app.use(
  bodyParser.urlencoded({
    extended: true
  })
);

app.get('/', (req, res) => {
  res.send('Hello World!')
})


app.post('/login', async (req, res) => {
  const uData = req.body;
  if (uData && uData.password && uData.email) {
    const regex = /^\S+@\S+\.\S+$/;
    if (regex.test(uData.email)) {
      const toMatch = {
        email: uData.email,
        password: md5(uData.password)
      }
      await DatabaseHelper.table('users').select(['uuid', 'username', 'email']).where(toMatch).then((data) => {
        if (data.length == 0) {
          res.status(401).send()
        }
        else {

          jwt = new jwtToken.TokenSigner('ES256K', "private key").sign(data[0])
          res.status(200).send(jwt)

        }
      })
        .catch((e) => {
          res.status(401).send(e)
        })
    }
    else {
      res.status(400).send()
    }
  }
  else {
    res.status(400).send()
  }
})



app.post('/register', async (req, res) => {
  const uData = req.body;
  if (uData && uData.username && uData.password && uData.email) {
    const regex = /^\S+@\S+\.\S+$/;
    if (regex.test(uData.email)) {
      // insert
      const toInsert = {
        uuid: UUIDHelper.generateUUID(),
        username: uData.username,
        email: uData.email,
        password: md5(uData.password)
      }
      await DatabaseHelper.table('users').insert(toInsert).then((data) => {
        res.status(200).send()
      })
        .catch((e) => {
          res.status(401).send(e)
        })
    }
    else {
      res.status(400).send()
    }
  }
  else {
    res.status(400).send()
  }
})



app.get('/join', async (req, res) => {
  await DatabaseHelper
    .table('items')
    .join('lists', DatabaseHelper.raw('item.uuid::varchar'), 'lists.item_id')
    .select('lists.*', 'items.*')
    .then((data) => {
      res.send(data)
    })

})

app.get('/questions', AuthHelper.tokenValidator, async (req, res) => {
  await DatabaseHelper.table('records').select('*').where({ user_id: req.body.user.uuid }).then((data) => {
    res.send(data);
  }).catch((error) => {
    res.send(error).status(400)
  })
})

app.get('/question/:uuid', AuthHelper.tokenValidator, async (req, res) => {
  if (req.params.uuid) {
    await DatabaseHelper.table('records').select('*').where({ uuid: req.params.uuid }).then((data) => {
      if (data.length > 0) {
        res.send(data[0]);
      }
      else {
        // could not find
        res.sendStatus(404)
      }
    }).catch((error) => {
      res.send(error).status(400)
    })
  }
  else {
    res.send(400)
  }
})
/**
* 
*/
app.post('/question', AuthHelper.tokenValidator, async (req, res) => {
  const question = req.body.question;
  const response = await ConversationHelpers.senseEmotionHelper(question)
  const uuid = UUIDHelper.generateUUID();
  if (response) {
    const toInsertQuestion = {
      uuid: uuid,
      question: question,
      answer: response.toString(),
      user_id: req.body.user.uuid
    }
    await DatabaseHelper.insert(toInsertQuestion).table('records').returning('*').then(async (data) => {
      if (response == null) {
        res.sendStatus(402)
      }
      else {
        const answer = { ...ConversationHelpers.convertEmotionValue(response), uuid: uuid };
        res.send(answer);
      }
    }).catch((e) => {
      res.status(401).send(e)
    })
  }
  else {
    res.status(400).send()
  }
})


app.patch('/question/:uuid', AuthHelper.tokenValidator, async (req, res) => {
  if (req.params.uuid && req.body) {
    const toAlter = {};
    if (req.body.question) {
      toAlter["question"] = req.body.question;
    }
    await DatabaseHelper.table('records').update(toAlter).where({ uuid: req.params.uuid }).returning('*').then((data) => {
      if (data.length > 0) {
        res.status(200).send(data[0]);
      }
      else {
        res.status(404).send();
      }
    }).catch((error) => {
      res.status(403).send(error)
    })
  }
  else {
    res.sendStatus(400)
  }
})
app.delete('/question/:uuid', AuthHelper.tokenValidator, async (req, res) => {
  if (req.params.uuid) {
    await DatabaseHelper.table('records').delete().where({ uuid: req.params.uuid }).returning('*').then((data) => {
      if (data.length > 0) {
        res.sendStatus(200);
      }
      else {
        res.sendStatus(404);
      }
    }).catch((error) => {
      res.send(error).status(400)
    })
  }
  else {
    res.send(400)
  }
})

if (process.env.NODE_ENV !== 'test') {
  app.listen(process.env.PORT || 3000, () => console.log(`Listening on port ${process.env.PORT || 3000}`));
}

module.exports = app